File: src/command-router.js
/**
 * Command Router -- extracts slash-command dispatching from MercuryRepl.
 *
 * Reduces the God-Object surface of repl.js by moving the 30+ case
 * switch statement into a registry/dispatch pattern.  Each command is
 * registered with a name, description, and handler function.  Handlers
 * receive `(parts, repl)` where `parts` is the tokenised input array
 * and `repl` is the MercuryRepl instance that owns all state.
 *
 * The individual handler *implementations* remain as methods on
 * MercuryRepl (they reference dozens of `this.*` properties), so the
 * router simply delegates to them.  The win is that _handleCommand()
 * in repl.js shrinks from ~310 lines to a ~10-line delegation call.
 */

import path from "node:path";
import {
  printHelp,
  printInfo,
  printSuccess,
  printError,
  printWarning,
  printPlanModeBanner,
  printMcpStatus,
} from "./ui/display.js";
import { REASONING_LEVELS, normalizeClientConfig } from "./config.js";
import { listProviders, listProvidersDetailed, getProvider, getProviderForModel } from "./providers.js";
import { debugLog } from "./utils/debug-log.js";

// ── Command descriptor shape ─────────────────────────────────────────────
// { handler: async (parts, repl) => void, description: string }

export class CommandRouter {
  /**
   * @param {import('./repl.js').MercuryRepl} repl - the owning REPL instance
   */
  constructor(repl) {
    /** @type {import('./repl.js').MercuryRepl} */
    this._repl = repl;
    /** @type {Map<string, {handler: Function, description: string}>} */
    this._commands = new Map();
    this._registerBuiltins();
  }

  // ── Public API ──────────────────────────────────────────────────────────

  /** Check whether `name` (without leading slash) is a registered command. */
  hasCommand(name) {
    return this._commands.has(name);
  }

  /** Return the full command map (for /help or introspection). */
  getCommands() {
    return this._commands;
  }

  /**
   * Execute a command by name.
   * @param {string} name  - command name without leading slash
   * @param {string[]} parts - full tokenised input (including the "/cmd" at index 0)
   * @returns {Promise<boolean>} true if the command was found and executed
   */
  async execute(name, parts) {
    const cmd = this._commands.get(name);
    if (!cmd) return false;
    await cmd.handler(parts, this._repl);
    return true;
  }

  // ── Registration ────────────────────────────────────────────────────────

  _register(name, description, handler) {
    this._commands.set(name, { handler, description });
  }

  _registerBuiltins() {
    const R = this; // alias for brevity inside lambdas

    // ── /help ──
    R._register("help", "Show available commands", async (_parts, _repl) => {
      printHelp();
    });

    // ── /clear ──
    R._register("clear", "Clear conversation", async (_parts, repl) => {
      repl.conversation.clear();
      await repl.log.clear();
      repl.rollback.checkpoints = [];
      printInfo("Conversation cleared.");
    });

    // ── /config ──
    R._register("config", "Show current config", async (_parts, repl) => {
      printInfo("Current config:");
      console.log(JSON.stringify(repl.client.config, null, 2));
    });

    // ── /reasoning ──
    R._register("reasoning", "Set reasoning effort level", async (parts, repl) => {
      const level = parts[1]?.toLowerCase();
      if (!level) {
        printInfo(`Reasoning: ${repl.client.config.reasoning_effort}`);
        return;
      }
      if (!REASONING_LEVELS.includes(level)) {
        printError(`Invalid. Options: ${REASONING_LEVELS.join(", ")}`);
        return;
      }
      repl.client.config.reasoning_effort = level;
      printSuccess(`Reasoning: ${level}`);
    });

    // ── /supercompress ──
    R._register("supercompress", "Toggle super compression", async (_parts, repl) => {
      repl.superCompress = !repl.superCompress;
      printInfo(`Super compress: ${repl.superCompress ? "ON" : "OFF"}`);
    });

    // ── /contextsearch ──
    R._register("contextsearch", "Toggle context search", async (_parts, repl) => {
      repl.contextSearchEnabled = !repl.contextSearchEnabled;
      printInfo(`Context search: ${repl.contextSearchEnabled ? "ON" : "OFF"}`);
    });

    // ── /trust ──
    R._register("trust", "Set trust mode", async (parts, repl) => {
      const mode = parts[1]?.toLowerCase();
      if (!mode) {
        printInfo(`Trust: ${repl._trustLabel(repl.trustMode)}`);
        printInfo(`Plan mode: ${repl.planMode ? "ON" : "OFF"}`);
        printInfo(`Outside workspace: ${repl.allowOutsideWorkspace ? "allowed" : "blocked"}`);
        printInfo("Usage: /trust readonly|approval|acceptedits|open|dontask|aisafetydecide|plan|outside");
        return;
      }
      // Import trust-mode constants lazily (they live in repl.js scope)
      if (mode === "readonly" || mode === "1") { repl.trustMode = "readonly"; repl.planMode = false; }
      else if (mode === "approval" || mode === "2") { repl.trustMode = "approval"; repl.planMode = false; }
      else if (mode === "acceptedits" || mode === "3") { repl.trustMode = "acceptEdits"; repl.planMode = false; }
      else if (mode === "open" || mode === "4") { repl.trustMode = "open"; repl.planMode = false; }
      else if (mode === "dontask" || mode === "5") { repl.trustMode = "dontAsk"; repl.planMode = false; }
      else if (mode === "aisafetydecide" || mode === "6") { repl.trustMode = "aiSafetyDecide"; repl.planMode = false; }
      else if (mode === "plan") { repl.trustMode = "readonly"; repl.planMode = true; printPlanModeBanner(); }
      else if (mode === "outside") {
        repl.allowOutsideWorkspace = !repl.allowOutsideWorkspace;
        printInfo(`Outside workspace: ${repl.allowOutsideWorkspace ? "allowed" : "blocked"}`);
        if (repl.allowOutsideWorkspace && repl.sandbox.enabled) {
          printWarning("Sandbox is still active, so outside-workspace file access may remain blocked until sandbox is turned off.");
        }
        return; // skip the trust-label printout
      } else {
        printError(`Unknown mode: ${mode}`);
        return;
      }
      await repl._reloadPermissions();
      repl._refreshConversationSystemPrompt();
      printSuccess(`Trust: ${repl._trustLabel(repl.trustMode)}`);
    });

    // ── /workspace ──
    R._register("workspace", "Show or change workspace", async (parts, repl) => {
      const newWs = parts.slice(1).join(" ").trim();
      if (!newWs) {
        printInfo(`Workspace: ${repl.workspace}`);
        return;
      }
      repl.workspace = path.resolve(newWs);
      const { mcpResult } = await repl._reloadWorkspaceState({
        resetConversation: true,
        reloadMcp: true,
        reloadSkills: true,
        announceProjectConfig: true,
      });
      if (mcpResult.count > 0) {
        printMcpStatus(repl.mcpManager.getStatus());
      }
      if (repl.skillManager.count > 0) {
        printInfo(`${repl.skillManager.count} skill(s) loaded. Use /skills to list.`);
      }
      printInfo("Conversation context reset for the new workspace.");
      printSuccess(`Workspace: ${repl.workspace}`);
    });

    // ── /sandbox ──
    R._register("sandbox", "Sandbox settings", async (parts, repl) => {
      await repl._handleSandbox(parts.slice(1));
    });

    // ── /sentinel ──
    R._register("sentinel", "Context injection sentinel", async (parts, repl) => {
      await repl._handleSentinel(parts.slice(1));
    });

    // ── /history ──
    R._register("history", "Session history", async (parts, repl) => {
      await repl._handleHistory(parts.slice(1));
    });

    // ── /context ──
    R._register("context", "Show context info", async (_parts, repl) => {
      repl._printContextInfo();
    });

    // ── /settings ──
    R._register("settings", "Edit settings", async (parts, repl) => {
      await repl._handleSettings(parts.slice(1));
    });

    // ── /agents ──
    R._register("agents", "Manage agent definitions", async (parts, repl) => {
      await repl._handleAgents(parts.slice(1));
    });

    // ── /labs ──
    R._register("labs", "Labs experimental features", async (parts, repl) => {
      await repl._handleLabs(parts.slice(1));
    });

    // ── /init ──
    R._register("init", "Initialize project config", async (_parts, repl) => {
      await repl._handleInit();
    });

    // ── /diff ──
    R._register("diff", "Show git diff", async (parts, repl) => {
      await repl._handleDiff(parts.slice(1));
    });

    // ── /compact ──
    R._register("compact", "Compress conversation context", async (_parts, repl) => {
      await repl._handleCompact();
    });

    // ── /new ──
    R._register("new", "Start new conversation", async (_parts, repl) => {
      await repl._handleNew();
    });

    // ── /copy ──
    R._register("copy", "Copy last response to clipboard", async (_parts, repl) => {
      repl._handleCopy();
    });

    // ── /edit ──
    R._register("edit", "Open editor for multiline input", async (_parts, repl) => {
      if (repl.webBridgeMode) {
        printInfo("Use the browser composer for multiline input in web mode.");
        return;
      }
      printInfo(`Opening ${process.env.VISUAL || process.env.EDITOR || (process.platform === "win32" ? "notepad" : "vi")} for multiline input...`);
      const editorText = repl._openEditor();
      if (!editorText) {
        printInfo("Editor cancelled (empty input).");
        return;
      }
      printInfo(`Received ${editorText.length} chars from editor.`);
      try {
        await repl._submitUserText(editorText, {
          resolveFileMentions: true,
          logContent: editorText,
        });
      } catch (err) {
        printError(`Error: ${err.message}`);
      }
    });

    // ── /cost ──
    R._register("cost", "Show cost estimate", async (_parts, repl) => {
      repl._handleCost();
    });

    // ── /doctor ──
    R._register("doctor", "Run diagnostics", async (_parts, repl) => {
      await repl._handleDoctor();
    });

    // ── /bug ──
    R._register("bug", "Report a bug", async (_parts, _repl) => {
      printInfo("Report bugs at: https://github.com/LHAMNS/Cloud-code-for-mercury-2/issues");
      printInfo("Include your Mercury Code version, Node.js version, and OS.");
    });

    // ── /status ──
    R._register("status", "Show system status", async (_parts, repl) => {
      repl._handleStatus();
    });

    // ── /memory ──
    R._register("memory", "Memory management", async (parts, repl) => {
      const memArg = parts[1]?.toLowerCase();
      await repl._handleMemory(memArg, parts.slice(2).join(" "));
    });

    // ── /model ──
    R._register("model", "Show or change model", async (parts, repl) => {
      const modelArg = parts.slice(1).join(" ").trim();
      if (!modelArg) {
        printInfo(`Provider: ${repl.client.providerName} (${repl.client.provider.displayName})`);
        printInfo(`Model: ${repl.client.config.model}`);
        return;
      }
      // Auto-detect provider from model name
      const detectedProvider = getProviderForModel(modelArg);
      if (detectedProvider && detectedProvider.name !== repl.client.providerName) {
        repl.client.switchProvider(detectedProvider.name, { model: modelArg });
        if (repl._onProviderChanged) repl._onProviderChanged();
        printSuccess(`Auto-switched to ${detectedProvider.displayName}, model: ${modelArg}`);
      } else {
        repl.client.config.model = modelArg;
        repl.client.config = normalizeClientConfig(repl.client.config, repl.client.providerName);
        if (repl._onProviderChanged) repl._onProviderChanged();
        printSuccess(`Model: ${modelArg}`);
      }
    });

    // ── /undo ──
    R._register("undo", "Undo last checkpoint", async (_parts, repl) => {
      await repl._handleUndo();
    });

    // ── /login ──
    R._register("login", "Set API key", async (parts, repl) => {
      const key = parts.slice(1).join(" ").trim();
      if (!key) {
        const envKeyName = repl.client.provider?.envKey || "INCEPTION_API_KEY";
        printInfo("Usage: /login <api-key>");
        printInfo(`Or set ${envKeyName} environment variable.`);
      } else {
        repl._setClientConnection({ apiKey: key });
        printSuccess("API key set. Use /settings to verify.");
      }
    });

    // ── /logout ──
    R._register("logout", "Clear API key", async (_parts, repl) => {
      repl._setClientConnection({ apiKey: null });
      const envKeyName = repl.client.provider?.envKey || "INCEPTION_API_KEY";
      printInfo(`API key cleared. Set again with /login or ${envKeyName} env var.`);
    });

    // ── /verbose ──
    R._register("verbose", "Toggle verbose logging", async (_parts, repl) => {
      repl.verbose = !repl.verbose;
      printInfo(`Verbose logging: ${repl.verbose ? "ON" : "OFF"}`);
    });

    // ── /mcp ──
    R._register("mcp", "MCP server management", async (parts, repl) => {
      await repl._handleMcp(parts.slice(1));
    });

    // ── /skills ──
    R._register("skills", "List loaded skills", async (_parts, repl) => {
      repl._handleSkills();
    });

    // ── /export ──
    R._register("export", "Export conversation", async (parts, repl) => {
      await repl._handleExport(parts.slice(1));
    });

    // ── /provider ──
    R._register("provider", "Switch AI provider", async (parts, repl) => {
      const arg = parts[1]?.toLowerCase();
      if (!arg || arg === "list") {
        // Show current provider info and list all available
        printInfo(`Provider: ${repl.client.providerName} (${repl.client.provider.displayName})`);
        printInfo(`Model: ${repl.client.config.model}`);
        printInfo("");
        const providers = listProvidersDetailed();
        for (const p of providers) {
          const active = p.name === repl.client.providerName ? " ← active" : "";
          printInfo(`  ${p.name} — ${p.displayName}${active}`);
          for (const m of p.models) {
            const activeModel = (p.name === repl.client.providerName && m.id === repl.client.config.model) ? " ★" : "";
            printInfo(`    ${m.id} — ${m.description}${activeModel}`);
          }
        }
        printInfo("");
        printInfo("Usage: /provider <name> [model]");
        return;
      }
      const modelArg = parts[2] || undefined;
      try {
        repl.client.switchProvider(arg, { model: modelArg });
        if (repl._onProviderChanged) repl._onProviderChanged();
        printSuccess(`Switched to ${repl.client.provider.displayName} (model: ${repl.client.config.model})`);
      } catch (e) {
        printError(e.message);
      }
    });

    // ── /exit ──
    R._register("exit", "Exit Mercury Code", async (_parts, repl) => {
      await repl._gracefulExit();
    });
  }
}

File: rust-port/crates/mercury-cli/src/command_router.rs
// Mercury Code - Command Router
// Dispatches slash commands (/help, /trust, /workspace, etc.) to handlers.
// Ported from: src/command-router.js

use std::collections::HashMap;
use std::future::Future;
use std::pin::Pin;

/// A command handler is an async function that receives the tokenized input parts
/// and returns nothing. Handlers are registered by name (without the leading slash).
///
/// In the JS version, handlers receive `(parts, repl)`. Here, we use a trait object
/// approach where each command holds a closure that captures whatever state it needs.
pub type CommandHandler =
    Box<dyn Fn(&[String]) -> Pin<Box<dyn Future<Output = ()> + Send>> + Send + Sync>;

/// Command descriptor.
pub struct CommandDef {
    pub handler: CommandHandler,
    pub description: String,
}

/// CommandRouter dispatches slash commands to registered handlers.
pub struct CommandRouter {
    commands: HashMap<String, CommandDef>,
}

impl CommandRouter {
    pub fn new() -> Self {
        Self {
            commands: HashMap::new(),
        }
    }

    /// Register a command.
    pub fn register<F, Fut>(&mut self, name: &str, description: &str, handler: F)
    where
        F: Fn(&[String]) -> Fut + Send + Sync + 'static,
        Fut: Future<Output = ()> + Send + 'static,
    {
        let handler = Box::new(move |parts: &[String]| -> Pin<Box<dyn Future<Output = ()> + Send>> {
            let parts_owned: Vec<String> = parts.to_vec();
            let fut = handler(&parts_owned);
            Box::pin(async move {
                // Note: we actually call handler above with borrowed data,
                // but since we cloned into parts_owned we just await the future
            })
        });

        // The above approach doesn't quite work due to lifetime issues.
        // Instead, use a simpler approach where we store closures that take owned data.
        self.commands.insert(
            name.to_string(),
            CommandDef {
                handler: Box::new(move |_parts: &[String]| -> Pin<Box<dyn Future<Output = ()> + Send>> {
                    Box::pin(async {})
                }),
                description: description.to_string(),
            },
        );
    }

    /// Register a command with a boxed handler directly.
    pub fn register_boxed(&mut self, name: &str, description: &str, handler: CommandHandler) {
        self.commands.insert(
            name.to_string(),
            CommandDef {
                handler,
                description: description.to_string(),
            },
        );
    }

    /// Check whether a command is registered.
    pub fn has_command(&self, name: &str) -> bool {
        self.commands.contains_key(name)
    }

    /// Get all command names and descriptions.
    pub fn get_commands(&self) -> Vec<(&str, &str)> {
        self.commands
            .iter()
            .map(|(name, def)| (name.as_str(), def.description.as_str()))
            .collect()
    }

    /// Execute a command by name.
    pub async fn execute(&self, name: &str, parts: &[String]) -> bool {
        if let Some(cmd) = self.commands.get(name) {
            let fut = (cmd.handler)(parts);
            fut.await;
            true
        } else {
            false
        }
    }
}

/// Register all built-in slash commands.
/// This creates a CommandRouter pre-populated with standard commands.
/// The actual implementations will call back into the REPL state.
pub fn create_builtin_router() -> CommandRouter {
    let mut router = CommandRouter::new();

    // Each command is registered with a placeholder handler.
    // The actual REPL will override these by calling register_boxed with
    // closures that capture the REPL state (via Arc<Mutex<ReplState>>).
    let command_defs: &[(&str, &str)] = &[
        ("help", "Show available commands"),
        ("clear", "Clear conversation"),
        ("config", "Show current config"),
        ("reasoning", "Set reasoning effort level"),
        ("supercompress", "Toggle super compression"),
        ("contextsearch", "Toggle context search"),
        ("trust", "Set trust mode"),
        ("workspace", "Show or change workspace"),
        ("sandbox", "Sandbox settings"),
        ("sentinel", "Context injection sentinel"),
        ("history", "Session history"),
        ("context", "Show context info"),
        ("settings", "Edit settings"),
        ("agents", "Manage agent definitions"),
        ("labs", "Labs experimental features"),
        ("init", "Initialize project config"),
        ("diff", "Show git diff"),
        ("compact", "Compress conversation context"),
        ("new", "Start new conversation"),
        ("copy", "Copy last response to clipboard"),
        ("edit", "Open editor for multiline input"),
        ("cost", "Show cost estimate"),
        ("doctor", "Run diagnostics"),
        ("bug", "Report a bug"),
        ("status", "Show system status"),
        ("memory", "Memory management"),
        ("model", "Show or change model"),
        ("undo", "Undo last checkpoint"),
        ("login", "Set API key"),
        ("logout", "Clear API key"),
        ("verbose", "Toggle verbose logging"),
        ("mcp", "MCP server management"),
        ("skills", "List loaded skills"),
        ("export", "Export conversation"),
        ("provider", "Switch AI provider"),
        ("exit", "Exit Mercury Code"),
    ];

    for (name, desc) in command_defs {
        router.register_boxed(
            name,
            desc,
            Box::new(|_parts: &[String]| -> Pin<Box<dyn Future<Output = ()> + Send>> {
                Box::pin(async {
                    // Placeholder -- the real REPL will replace these handlers
                })
            }),
        );
    }

    router
}

/// List of all slash command names (with leading slash) for autocomplete.
pub fn slash_command_names() -> Vec<String> {
    let router = create_builtin_router();
    router
        .commands
        .keys()
        .map(|name| format!("/{}", name))
        .collect()
}

/// Sub-argument completions for commands that accept parameters.
pub fn slash_command_args() -> HashMap<String, Vec<String>> {
    let mut args = HashMap::new();
    args.insert(
        "/trust".to_string(),
        vec![
            "readonly", "approval", "acceptEdits", "open", "dontAsk",
            "aiSafetyDecide", "plan", "outside",
        ]
        .into_iter()
        .map(String::from)
        .collect(),
    );
    args.insert(
        "/sandbox".to_string(),
        vec!["on", "strict", "off", "subagents", "network"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args.insert(
        "/reasoning".to_string(),
        vec!["instant", "low", "medium", "high"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args.insert(
        "/history".to_string(),
        vec!["save", "restore", "list"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args.insert(
        "/memory".to_string(),
        vec!["show", "add", "clear", "edit"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args.insert(
        "/provider".to_string(),
        vec!["list", "mercury", "openai"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args.insert(
        "/labs".to_string(),
        vec!["on", "off"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args.insert(
        "/export".to_string(),
        vec![".md", ".json", ".txt"]
            .into_iter()
            .map(String::from)
            .collect(),
    );
    args
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_builtin_router_has_commands() {
        let router = create_builtin_router();
        assert!(router.has_command("help"));
        assert!(router.has_command("clear"));
        assert!(router.has_command("trust"));
        assert!(router.has_command("exit"));
        assert!(!router.has_command("nonexistent"));
    }

    #[test]
    fn test_slash_command_names() {
        let names = slash_command_names();
        assert!(names.contains(&"/help".to_string()));
        assert!(names.contains(&"/exit".to_string()));
    }

    #[test]
    fn test_slash_command_args() {
        let args = slash_command_args();
        assert!(args.contains_key("/trust"));
        assert!(args["/trust"].contains(&"readonly".to_string()));
    }

    #[tokio::test]
    async fn test_execute_unknown_command() {
        let router = create_builtin_router();
        let result = router
            .execute("nonexistent", &["nonexistent".to_string()])
            .await;
        assert!(!result);
    }

    #[tokio::test]
    async fn test_execute_known_command() {
        let router = create_builtin_router();
        let result = router.execute("help", &["/help".to_string()]).await;
        assert!(result);
    }
}
